File: 哈希表/最大相等频率.java
package 哈希表;

import java.util.HashMap;


public class 最大相等频率 {
    int res = 0;
    int maxFre = 0; //记录当前的最大频率
    HashMap<Integer, Integer> count = new HashMap<Integer, Integer>();
    HashMap<Integer, Integer> fre = new HashMap<Integer, Integer>();

    public int maxEqualFreq(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            if (!count.containsKey(nums[i])) {
                count.put(nums[i], 1);
                fre.put(1, fre.getOrDefault(1, 0) + 1);
                maxFre = Math.max(maxFre, 1);
            } else {
                fre.put(count.get(nums[i]), fre.get(count.get(nums[i])) - 1);
                count.put(nums[i], count.get(nums[i]) + 1);
                maxFre = Math.max(maxFre,count.get(nums[i]));
                fre.put(count.get(nums[i]), fre.getOrDefault(count.get(nums[i]), 0) + 1); //等价于go中的fre[count[nums[i]]]++
            }
            boolean ok = maxFre == 1 ||
                    fre.get(maxFre) * maxFre + fre.get(maxFre - 1) * (maxFre - 1) == i + 1 && fre.get(maxFre) == 1 ||
                    fre.get(1) + fre.get(maxFre) * maxFre == i + 1 && fre.get(1) == 1;

            if (ok) {
                res = i + 1;
            }
        }

        return res;
    }
}

File: 哈希表/最大相等频率.go
package 哈希表

func maxEqualFreq(nums []int) int {
	maxFre := 0            //维护遍历数组过程中，如果一个数字他的频率最大，那么这个数字的频率就为maxFre
	count := map[int]int{} //数字i出现的频率为count[i]
	fre := map[int]int{}   //出现频率为i的数字有fre[i]个
	res := 0
	for i := 0; i < len(nums); i++ { //当前一共有i+1个数字
		if _, ok := count[nums[i]]; !ok { //目前还不存在数字nums[i]
			count[nums[i]] = 1 //nums[i]出现的次数现在为1
			fre[1]++           //频率为1的数字增加了一个
			maxFre = max(maxFre, 1)
		} else {
			fre[count[nums[i]]]--
			count[nums[i]]++
			fre[count[nums[i]]]++
			maxFre = max(maxFre, count[nums[i]])
		}
		if maxFre == 1 ||
			fre[maxFre]*maxFre+fre[maxFre-1]*(maxFre-1) == i+1 && fre[maxFre] == 1 ||
			fre[1]+fre[maxFre]*maxFre == i+1 && fre[1] == 1 {
			res = i + 1
		}
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}
